Color Picker: improve UX when dragging the handle on top of editor canvas

ciampo opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Fixes #49267

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

As explained in #49267, currently when a `ColorPicker` is rendered inside a `Popover`:

- the handle is cutoff
- dragging the handle doesn't work when the cursor leaves the picker's boundaries

This creates a sub-optimal user experience when using the color picker in the editor.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

- Tweaking some `overflow` CSS properties
- Disabling the `resize` option for the `Popover` when rendering the `ColorPicker` (which was also applying some unwanted `overflow` rules)

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->

TBD

## Screenshots or screencast <!-- if applicable -->

ciampo committed:
Color Picker: force overflow visible when inside a popover

ciampo commented:
@jameskoster in my tests, this approach seems to show the handle, and also allows the handle to be dragged around when the cursor is moved beyond the boundaries of the popover.

Unfortunately, though, in my tests, the handle jumps around instead of moving smoothly when dragging the cursor outside of the popover. Even worse, after some amount of dragging, the popover with the color picker closes unexpectedly

Here's a video to show what I mean:


https://github.com/WordPress/gutenberg/assets/1083581/cb3c058d-d143-4f27-b0f9-b0570932b374


As I suspected, the reason for the handle "jitteriness" and the unexpected popover closure could be related to the fact that the pointer-events happen on top of an iframe — I tried quickly hiding the iframe in the dev tools and the dragging experience improved immediately.


I treated this as a timeboxed effort on my end, and I don't really currently have much more time to look into this.

ciampo committed:
Disable `resize` on popover instead of hacky CSS snippet

jameskoster commented:
That's a shame. 

I could live with the slight mis-positioning of the handle as it is still a usability improvement on trunk imo, but the intermittent disappearing act is going to be way too frustrating.

Thanks for giving it a shot.

ciampo committed:
Update react-colorful to version 5.6.1

ciampo committed:
Add test story with iframe

ciampo committed:
Attempt fix by disabling iframe pointer events

ciampo commented:
I spent some extra time working on this, and I can confirm that the issue is caused by the underlying `iframe` (the editor canvas, in this case). That's because an `iframe` is technically another window, and therefore it "absorbs" pointer events.

I put together a bit of a hacky workaround: when the user starts to drag the pointer on the saturation picker, we search for the `iframe` and we disable pointer events on it. The results seem to be promising:

https://github.com/WordPress/gutenberg/assets/1083581/6bb033c7-8300-4c8e-8fcb-fb2c5075f13d

Now, the next steps could be to refine this approach into a more robust solution. Here's what I had in mind:

- tweak the `iFrame` component to be in a "draggable" state — that can be done by ether disabling pointer events, or by rendering a transparent pseudo element that captures pointer events.- have `ColorPicker` expose a couple of callbacks (something like `onDragStrart` and `onDragEnd`)
- expand the dragging events not only of the 2d picker, but also for pointer events on any slider / input
- in the block editor, pass callbacks to `ColorPicker` that would be able to toggle the `iFrame`'s draggable state

We'll also need to be careful because, if for any reason the code isn't able to restore pointer events on the iframe, the whole canvas would become impossible to interact with for our users.

Pinging a few folks to see if the approach highlighted above seems good at least on paper (@youknowriad @jsnajdr @tyxla )

ciampo commented on packages/components/src/color-picker/picker.tsx:
```diff
@@ -19,6 +19,46 @@ export const Picker = ( { color, enableAlpha, onChange }: PickerProps ) => {
+		// issue from happening.
+		const disableIframePointerEvents = () => {
+			iframe.style.pointerEvents = 'none';
+			window.addEventListener( 'pointerup', restoreIframePointerEvents );
```
Using `window` for the `pointerup` event because we'd need to make sure that pointer events on the iframe are restored whenever the pointer is released

youknowriad commented:
@ciampo In the Iframe component, we do have a `bubbleEvents` function that we use to bubble event from the iframe to the parent document. Do you think that adding the events that we need for this bug there would be an alternative solution? I have nothing against the current solution but wanted to check if this might be a "common issue" that was already solved for other interactions?

ciampo commented:
> @ciampo In the Iframe component, we do have a `bubbleEvents` function that we use to bubble event from the iframe to the parent document. Do you think that adding the events that we need for this bug there would be an alternative solution? I have nothing against the current solution but wanted to check if this might be a "common issue" that was already solved for other interactions?

I wasn't aware of that, but it makes sense — without the event bubbling, I guess the color picker's handle wouldn't move at all, while it currently moves (albeit in a jittery and buggy way).

But unfortunately, it still doesn't seem to fix the problem with the popover closing unexpectedly. Somehow, dragging the pointer around while tweaking the color picker's color somehow causes the `iframe` to receive focus (not sure exactly why) — but the loss of focus causes the dropdown to close abruptly. That's why the solution that I tested above (disabling pointer events) seems to work well, since it prevents the iframe from receiving focus and lets the browser capture `pointermove` events natively (ie. without forwarding them across the iframe).

I'll see if I can work out a solution. Regarding this part:

> in the block editor, pass callbacks to ColorPicker that would be able to toggle the iFrame's draggable state

I was wondering what would the best way be for components across the editor to toggle the iframe not reacting to pointer events — maybe using React context?

youknowriad commented:
I think @ellatrix would be a good one to involve in this discussion

ciampo committed:
ColorPicker: refactor component to take a drag start and drag end props (prop name TBD)

ciampo committed:
Forward the drag start/end prop on all affected components in the package

ciampo committed:
comment idea

ciampo committed:
Add block canvas context, use across the post editor

ciampo committed:
Add test onPopoverClose prop to avoid the iframe being stuck when the popover closes unexpectedly

ciampo commented:
I went ahead and started to implement the solution described in https://github.com/WordPress/gutenberg/pull/54164#issuecomment-1740525188 to get a better feeling for it:

- added props to `ColorPicker` and all components in `@wordpress/components` using `ColorPicker` internally, exposing callbacks for when the dragging starts and ends
- added a new `BlockCanvasContext` to allow different parts of the editor to enable/disable pointer events on the editor canvas
- tested the new approach for the sidebar controls that allow setting text/background/link colors. It seems to be working as expected
- added an additional callback for when the popover containing the color picker closes before the drag ends

The approach seems to work, although:
- it's a bit clunky (we'll need to manually apply the fix for every usage of color picker in the editor)
- i haven't tested it in depth yet, but in case the logic ever had an edge case that is not covered, we'd risk making the editor canvas non-interactive (quite a bad regression)

I guess I'd like to discuss with @ellatrix if there are any better approaches to this — for example, I'd like to understand why is the iframe currently gaining focus while dragging the mouse around.

ellatrix commented:
Probably the iframe focus is native browser behaviour, I don't think we're explicitly setting focus on drag over.

Maybe disable pointer events for everything outside the color picker while dragging?

ciampo commented:
Testing an alternative approach in https://github.com/WordPress/gutenberg/pull/55149

ciampo commented:
Closing in favour of https://github.com/WordPress/gutenberg/pull/55149
